Add script class hierarchies & add-script button permanence/auto-derivation

willnationsdev opened the pull request:
Previously, if you attempted to create a script class that extends another script, it wouldn't show up in the CreateDialog. I have altered the backend to fix this.

This PR allows you to have hierarchies of script classes that show up in the CreateDialog. It also modifies the scene tree dock so that the Add Script button is the only one that exists (no remove button - it's just hidden) and the add script button will auto-populate the Inherits field of the ScriptCreateDialog if the selected node already has a script. Finally, I made it so that script classes that are part of an editor plugin will be added/removed from the CreateDialog when their respective EditorPlugin is toggled active/inactive.

In the example below, "new_script.gd" is in a plugin I created, and you can see the effect on the Derived class in my res:// folder which derives it whenever you toggle the plugin.

Plugin Active:
![create_dialog1](https://user-images.githubusercontent.com/16217563/42847822-e0e06ec4-89e2-11e8-9711-5ad5e80ea27a.png)

Plugin Inactive:
![create_dialog2](https://user-images.githubusercontent.com/16217563/42847837-edd8ed7c-89e2-11e8-9234-d70fbf8a6068.png)

willnationsdev commented:
I can follow this PR up with separate PRs to add custom icons and permit GDScript inheritance via class name, whenever we decide when/how we wanna do that @reduz.

henriiquecampos commented:
Now we are talking!

My concern is: if I inherit using relative paths will this work as well?

E.g:

`extends "../foo/bar.gd"`
instead of 
`extends "res://baz/foo/bar.gd"`

Or this doesn't affect at all this feature? :thinking:

willnationsdev commented:
@henriiquecampos Inheritance via relative or absolute paths has never been a problem for GDScript...? Shouldn't have an impact here.

The thing I might follow this up with (if it doesn't work) is letting you inherit by class name. I tried it earlier, but it wasn't working, don't know yet if any changes I've made would fix that (since I had to modify the backend quite a bit to add this feature).

willnationsdev commented:
@vnen sounds good

willnationsdev commented:
@henriiquecampos oh interesting. I didn't even know that. That sounds like it would be fixable though...

mhilbrunner commented:
As discussed in the IRC meeting, the idea behind this is good and welcome, but needs to be implemented in a solid way.

Notes from IRC:
```
reduz> I still think bojidar_bgs PR for script categories was the best solution to this problem
too bad it was closed
the approach was bad, but the idea was good
a keyword export_category "Something" would be great
Akien> I think we're starting to have a need for annotations support in GDScript finally :)
Akien>We could do stuff like:
#[category "Something"]
export var foo
```

mhilbrunner commented:
D'oh. Thanks!

willnationsdev commented:
@mhilbrunner This one should be all good to go too.

willnationsdev commented:
I updated this one so that there is a ProjectSetting it checks for using a script class's name for inheritance rather than its filepath. I also fixed a bug that prevented the created script from opening in the ScriptEditor immediately after creation.

willnationsdev committed:
Add script hierarchies,add-script btn auto-derives

willnationsdev commented:
I've refactored this so that it makes no core changes. Purely CreateDialog, SceneTreeDock, and EditorData.

mhilbrunner commented:
Thanks for this contribution, by the way :)

OliPerraul commented:
I am confused, since when can we add scripts directly in the scene tree like that? This pr concerns inheritance in the tree but not the base feature.

willnationsdev commented:
@OliPerraul reduz added this feature about a week ago with commit a3f1ee5. If you use "class_name" in a GDScript, it now registers it as a global in the ScriptServer class. The CreateDialog was also edited to automatically display these classes, however the way it was done did now allow for inheritance between scripts to be demonstrated in the CreateDialog.

OliPerraul commented:
OK thanks

OliPerraul commented:
Is this the proper syntax
```
class_name CustomClass0
extends Control

func _ready():
	pass




```

willnationsdev commented:
@OliPerraul No, the `class_name` keyword must be placed after the extends statement, as far as I know.
